docs/docs/guides/03-typescript.md

```diff
@@ -2,4 +2,266 @@
+
+// src/index.ts
+import type { Service, UID } from './index';
+
+type AdminService = Record<UID.Admin, Service.Admin>;
+```
+
+### b. Imports
+
+When importing **only types** from a module, `import type` must be used.
+
+The `import type` syntax imports only the types and eliminates the runtime impact of importing the actual values. This helps reduce the bundle size and improves performance by avoiding unnecessary code execution.
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { Attribute, Common } from '@strapi/strapi';
+
+// ✅ do
+
+import type { Attribute, Common } from '@strapi/strapi';
+```
+
+When importing **both types and implementations** from a module:
+
+- `import type` must be used for types
+- `import` must be used for the rest
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { User, findUser } from './user';
+
+const user: User = findUser();
+
+// ✅ do
+
+import type { User } from './user';
+import { findUser } from './user';
+
+const user: User = findUser();
+```
+
+### c. Exports
+
+In most cases, it is recommended to use named exports (`export`) instead of the default export (`export default`). Named exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
+
+However, there are a few exceptions where using export default might be appropriate. For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.
+
+Using regular exports provides a clearer and more explicit interface for module consumers, promoting better code organization and maintainability. However, it is important to consider the context before deciding whether to use default exports.
+
+## 3. Parameters for Generics
+
+When working with generic types, it is required to follow naming conventions and use descriptive names for type parameters. This helps improve code readability and maintainability over the whole codebase.
+
+- For simple types with a single self-explanatory parameter, it is allowed to **use `T`**
+- For any other generic types' parameters, it is required to **use a meaningful PascalCased variable name prefixed by `T`**
+
+Example:
+
+```typescript
+// ❌ don't
+function query<T extends object, U extends keyof T>(data: T, properties: U[]): { [K in U]: T[K] } {
+  // ...
+}
+
+// ✅ do
+function query<TValue extends object, TKey extends keyof TValue>(
+  data: TValue,
+  properties: TKey[]
+): { [TIter in TKey]: TValue[TIter] } {
+  // ...
+}
+
+// ✅ ok
+type NonEmpty<TValue extends string> = TValue extends '' ? never : TValue;
+
+// ✅ prefer
+type NonEmpty<T extends string> = T extends '' ? never : T;
+```
+
+Choose meaningful names for type parameters that accurately describe their purpose within the context of the function or type.
+
+## 4. Avoiding Breaking Changes for Users
+
```

Comment:
We can add a callout to verify if a type you change is exported by the package. (I'm wondering how much we could automate that eventually, there might be lib that will check that)